autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py

```diff
@@ -739,115 +743,72 @@ def load(cls, path, file_prefix="", reset_paths=False, verbose=True):
         obj.model.load_parameters(path + cls.params_file_name, ctx=obj.ctx)
         obj.summary_writer = None
         return obj
 
     def hyperparameter_tune(self, X_train, X_test, Y_train, Y_test, scheduler_options, **kwargs):
+        time_start = time.time()
         """ Performs HPO and sets self.params to best hyperparameter values """
         self.verbosity = kwargs.get('verbosity', 2)
         logger.log(15, "Beginning hyperparameter tuning for Neural Network...")
         self._set_default_searchspace() # changes non-specified default hyperparams from fixed values to search-spaces.
         if self.feature_types_metadata is None:
             raise ValueError("Trainer class must set feature_types_metadata for this model")
-        scheduler_func = scheduler_options[0] # Unpack tuple
+        scheduler_func = scheduler_options[0]
         scheduler_options = scheduler_options[1]
         if scheduler_func is None or scheduler_options is None:
             raise ValueError("scheduler_func and scheduler_options cannot be None for hyperparameter tuning")
         num_cpus = scheduler_options['resource']['num_cpus']
-        num_gpus = scheduler_options['resource']['num_gpus']
-        self.params['num_dataloading_workers'] = max(1, int(num_cpus/2.0))
-        if num_gpus >= 1:
-            self.params['ctx'] = mx.gpu() # Currently cannot use more than 1 GPU until scheduler works
-        else:
-            self.params['ctx'] = mx.cpu()
-        # self.params['ctx'] = mx.cpu() # use this in case embedding layer complains during predict() for HPO with GPU
+        # num_gpus = scheduler_options['resource']['num_gpus']  # TODO: Currently unused
 
-        start_time = time.time()
-        X_train = self.preprocess(X_train)
-        if self.features is None:
-            self.features = list(X_train.columns)
         params_copy = self.params.copy()
+
+        self.num_dataloading_workers = max(1, int(num_cpus/2.0))
+        self.batch_size = params_copy['batch_size']
+        train_dataset, test_dataset = self.generate_datasets(X_train=X_train, y_train=Y_train, params=params_copy, X_test=X_test, y_test=Y_test)
+        train_path = self.path + "train"
+        test_path = self.path + "validation"
+        train_dataset.save(file_prefix=train_path)
+        test_dataset.save(file_prefix=test_path)
+
         if not np.any([isinstance(params_copy[hyperparam], Space) for hyperparam in params_copy]):
             logger.warning("Warning: Attempting to do hyperparameter optimization without any search space (all hyperparameters are already fixed values)")
         else:
             logger.log(15, "Hyperparameter search space for Neural Network: ")
             for hyperparam in params_copy:
                 if isinstance(params_copy[hyperparam], Space):
                     logger.log(15, str(hyperparam)+ ":   "+str(params_copy[hyperparam]))
-        directory = self.path # path to directory where all remote workers store things
```

Comment:
This will be needed for distributed setting, so at least keep a TODO here.  According to Hang, the working directory on remote may be different when `upload_files()` is invoked.  The current distributed usage of autogluon (eg. for image classification) does not utilize dask's `upload_files()`, but instead requires user to manually setup all worker machines with identical file-structure + dataset, but this means preprocessing will be redundantly executed on all of them.

Reply:
Added todo